### ISSUE TITLE
avrdude 6.3-20171130 fails with attiny13a

### ISSUE THREAD

**Issue by bratoff** [arduino/Arduino#8198]
I receive the following error output when attempting to program an attiny13a using avrdude 6.3-20171130, which is installed by Arduino 1.8.7.  If I revert to the previous version of avrdude (by installing version 1.6.21 of Arduino AVR Boards), everything works normally.

```
avrdude: Version 6.3-20171130
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "C:\Program Files (x86)\Arduino\hardware\tools\avr/etc/avrdude.conf"

         Using Port                    : usb
         Using Programmer              : usbtiny
avrdude: usbdev_open(): Found USBtinyISP, bus:device: bus-0:\\.\libusb0-0001--0x1781-0x0c9f
         AVR Part                      : ATtiny13
         Chip Erase delay              : 4000 us
         PAGEL                         : P00
         BS2                           : P00
         RESET disposition             : dedicated
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65     5     4    0 no         64    4      0  4000  4000 0xff 0xff
           flash         65     6    32    0 yes      1024   32     32  4500  4500 0xff 0xff
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00
           lock           0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           calibration    0     0     0    0 no          2    0      0     0     0 0x00 0x00
           lfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00

         Programmer Type : USBtiny
         Description     : USBtiny simple USB programmer, https://learn.adafruit.com/usbtinyisp
avrdude: programmer operation not supported

avrdude: Using SCK period of 10 usec
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0xffffff (probably .avr8x_mega) (retrying)

Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0xffffff (probably .avr8x_mega) (retrying)

An error occurred while uploading the sketch
Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0xffffff (probably .avr8x_mega)
avrdude: Yikes!  Invalid device signature.
         Double check connections and try again, or use -F to override
         this check.


avrdude done.  Thank you.
```
I can reproduce the exact same behavior by running avrdude from a command prompt, completely outside of the Arduino IDE, using the same configuration file, command parameters and programmer.  IMO, this isolates the issue to the 20171130 release of avrdude.

I have tried several other AVR family chips, and of those I've tried only the attiny13a appears to have this problem.

I have also tried several different versions of avrdude.conf, and that made no difference.

I do not have any other type of programmer than the usbtiny, so I cannot determine if that is a factor.

**Comment by facchinm** [comment]
The first signal that something is going wrong seems `avrdude: programmer operation not supported` that comes from [here](https://github.com/facchinm/avrdude/blob/b6fc55323b48d989af8e83384dcb02739303129a/pgm.c), so it looks like the programmer is not being recognized.
Can you share the commandline you are using?

**Comment by matthijskooijman** [comment]
@facchinm, sounds like the programmer *is* supported, but doesn't implement everything that is being used (e.g. perhaps not block writes, no chipe erase or no voltage control, stuff like that). Still, if some operation that is expected is not executed, that could cause a problem. Pity the error message doesn't say what operation exactly.

**Comment by facchinm** [comment]
@bratoff we switched version to support new Microchip AVR chips with UPDI programming interface. The change should not affect older chips/programmers so I'll double check your commandline.
If you need the old avrdude back, either downgrade avr core to 1.6.21 (last version with old avrdude) or replace the binary with this one http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino9-i686-w64-mingw32.zip

**Comment by bratoff** [comment]
I have both versions of avrdude on my machine.  Rather than downgrade my Arduino configuration, I am now programming the atttiny13a from the command line so that I can force the older version of avrdude.
It would be a far cleaner and more convenient solution if there were a single version of avrdude that supported both the attiny13a and the newer UPDI chips.  Is it possible that whatever was done in avrdude to detect the new chips affected the older chips somehow?  Could there possibly be a change in avrdude.conf that would work around the problem?

**Comment by facchinm** [comment]
@bratoff the best way would be to bisect the avrdude history until we find the root cause, but unfortunately I don't have an usbtiny programmer to test it. Could you send both the successful output and the broken one launched with `-vvv` (superverbose) command line switch? Probably there's a hint on what is going on. Thanks!

**Comment by bratoff** [comment]
@facchinm Here are the console outputs from the two runs.  The two versions are definitely issuing different commands to initialize the chip and read the signature.

This is the successful run using the avrdude from the 1.6.21 AVR core:
[Test1.txt](https://github.com/arduino/Arduino/files/2597967/Test1.txt)

And this is the broken run using the avrdude from the latest AVR core:
[Test2.txt](https://github.com/arduino/Arduino/files/2597968/Test2.txt)

Thanks very much for your follow up!

**Comment by facchinm** [comment]
The only commit regarding `usbtiny` alone in the period between the two builds of avrdude is https://github.com/facchinm/avrdude/commit/5beac69ff87ee3a9743913600fc5259e541b3ae5 . I'm attaching a build with that commit reverted, if you can test it it would help a lot! 
[avrdude-6.3.0-arduino14-i686-w64-mingw32.zip](https://github.com/arduino/Arduino/files/2598917/avrdude-6.3.0-arduino14-i686-w64-mingw32.zip)

**Comment by facchinm** [comment]
Great to hear that! I think the patch can be further reduced into changing [this](https://github.com/facchinm/avrdude/commit/5beac69ff87ee3a9743913600fc5259e541b3ae5#diff-228209822e927366780c773e187fb14bR506) line; I'm sending you another binary for testing and if it works I'll file a PR for mainline avrdude.
```patch
diff --git a/usbtiny.c b/usbtiny.c
index 1f86b1d..8c303de 100644
--- a/usbtiny.c
+++ b/usbtiny.c
@@ -503,7 +503,7 @@ static int usbtiny_initialize (PROGRAMMER *pgm, AVRPART *p )
   }
 
   for (tries = 0; tries < 4; ++tries) {
-    if (pgm->program_enable(pgm, p) >= 0)
+    if (pgm->program_enable(pgm, p) == 0)
       break;
     // no response, RESET and try again
     if (usb_control(pgm, USBTINY_POWERUP,
-- 
```
[avrdude-6.3.0-arduino14-i686-w64-mingw32.zip](https://github.com/arduino/Arduino/files/2603393/avrdude-6.3.0-arduino14-i686-w64-mingw32.zip)

**Comment by bratoff** [comment]
@facchinm    Success!  Looks like that's the fix.  Here's the console output:
[Test4.txt](https://github.com/arduino/Arduino/files/2606073/Test4.txt)

Thanks again for your help.

**Comment by matthijskooijman** [comment]
@facchinm, Nice fix, some good digging there :-)

**Comment by bratoff** [comment]
Confirmed fix in Arduino release 1.8.9

**Comment by bratoff** [comment]
The avrdude team accepted the patch in January, but it has not been included in the 1.8.9 release.

**Comment by facchinm** [comment]
The patch has been merged in https://github.com/facchinm/avrdude/commit/a45d20332a63fb529b5e8c89c7fea2bdf914e809 and is available on megaAVR core since 1.8.1 (on AVR it will arrive soon).
Closing the issue as fixed
